Enhancement - fixed parsing of usernames and domains(src+dst) in panw

marioschaefer opened the pull request:
fixed parsing of usernames and domains(src+dst) to allow shortened domains("nt-auth...t"), domains that only contain a "." and usernames that contain ":"

I've also integrated the changes propsed by #8780, since these PR seems to be stalled.

<!-- Type of change
Please label this PR with one of the following labels, depending on the scope of your change:
- Bug
- Enhancement
- Breaking change
- Deprecation
-->

## Proposed commit message
fixed parsing of usernames and domains(src+dst) to allow shortened domains ("nt-auth...t\username"), domains that only contain a "." (".\username") and usernames that contain ":" ("x-forwarded-for: 10.1.1.1")

<!-- Mandatory
Explain here the changes you made on the PR.

Please explain:

- WHAT: patterns used, algorithms implemented, design architecture, message processing, etc.
- WHY:  the rationale/motivation for the changes

This text will be pasted into the squash dialog when the change is committed and will be
a long term historical record of the change to help future contributors understand the
change, please help them by making it clear and comprehensive, they may be you.

If the commit title is adequate to describe both of these things, The text here may be omitted
or replaced with "See title". The title of the PR will be used as the commit message title when
the merge is made and the "See title" marker will be removed if present.

The text here and the PR title will be subject to the PR review process.
-->

## Checklist

- [x] I have reviewed [tips for building integrations](https://github.com/elastic/integrations/blob/main/docs/tips_for_building_integrations.md) and this pull request is aligned with them.
- [x] I have verified that all data streams collect metrics or logs.
- [x] I have added an entry to my package's `changelog.yml` file.
- [x] I have verified that Kibana version constraints are current according to [guidelines](https://github.com/elastic/elastic-package/blob/master/docs/howto/stack_version_support.md#when-to-update-the-condition).

## Author's Checklist

<!-- Recommended
Add a checklist of things that are required to be reviewed in order to have the PR approved
-->
- [ ] 

## How to test this PR locally

<!-- Recommended
Explain here how this PR will be tested by the reviewer: commands, dependencies, steps, etc.
-->
just use the provided tests in `panw\data_stream\panos\_dev\test\pipeline\test-panw-panos-threat-sample.log`

## Related issues

<!-- Recommended
Link related issues below. Insert the issue link or reference after the word "Closes" if merging this should automatically close it.

- Closes #123
- Relates #123
- Requires #123
- Supersedes #123
-->
- Closes #8780 

## Screenshots

<!-- Optional
Add here screenshots presenting:
- Kibana UI forms presenting configuration options exposed by the integration
- dashboards with collected metrics or logs
-->

marioschaefer commented:
> ❌ Author of the following commits did not sign a [Contributor Agreement](https://www.elastic.co/contributor-agreement): [c7cadd5](https://github.com/elastic/integrations/commit/c7cadd5241b91d92781f2042f3d75f9cafa8afd2), [b18bf96](https://github.com/elastic/integrations/commit/b18bf96ac50e1f1a6195b3fe7450fe37adff2557), [9b1b176](https://github.com/elastic/integrations/commit/9b1b1760ad31ffebc02456f5ba91630e6027b100)
> 
> Please, read and sign the above mentioned agreement if you want to contribute to this project

can someone tell my why it's still in the status "not signed"? 
I've already signed the agreement and downloaded the PDF version of it.

marioschaefer commented:
I've noticed another problem:  
The Syslogformat of the Palo Alto Config Logs can be different, depending on the configuration of the device.  
see: [Config Log Fields](https://docs.paloaltonetworks.com/pan-os/11-1/pan-os-admin/monitoring/use-syslog-for-monitoring/syslog-field-descriptions/config-log-fields)
Notice the description of the fields named `before-change-detail` and `after-change-detail`.  
That means that there can be either 19 or 21 fields in the syslogline of a config-log.  
Since the CSV-Processor doesn't fail if the fields supplied under `target_fields` doesn't match the value count in the sourcefield supplied by `field`, i can't use the `on_failure` method of the processors.  
To work around this problem, I've used a script to split the sourcefield by "," with the painless `splitOnToken` method and write the `length` attribute of the returned array into a new field called `messagesplitlength` in the document.  
Then I'm using the `if` method of the processor to decide which processor to run, depending on the value of the new field.  
After that, I remove the new field `messagesplitlength` since I don't need that anymore.  
This is working fine under perfect conditions.  

Challenge:  
since i only split the string by "," and don't do any csv parsing, it can happen that i count the number of values incorrectly.  
Since a valid csv can also look like this:
'asdf,asdf,asdf' = 3 values with splitontoken and 3 valid csv-values
'asdf,"asdf,asdf,asdf",asdf' = 5 values with splitontoken - but actually/correctly only 3 valid csv-values  
I can imagine some situations in which this can happen, for example if someone add a comment to a commit that includes a ",".  
Or a "," in the description of an address-object, etc. etc. etc.  

alternative solution:
i use the csv processor with the 21-field layout to fill temporary fields.  
then i check whether something was actually written in the last field (this would mean that the two optional fields were actually filled).  
if the last field was filled correctly, i start another csv processor with the 21-field layout and write in the correct fields.  
However, if the last field is empty, I run the csv processor with the 19-field layout and then write the correct fields. 

My Question is now:  
Is there a better solution to this problem which I didn't see?

elasticmachine commented:
Pinging @elastic/sec-deployment-and-devices (Team:Security-Deployment and Devices)

taylor-swanson reviewed (changes requested):
Thank you for the contribution, @marioschaefer!

There are a few too many things happening in this PR and I recommend the following to help cut it down in size:

Let's keep the following changes since that seems to be the main goal of the PR, in particular:

- Parsing domains that only contain a '.' and usernames that only contain a ':'
- Parsing of truncated usernames and domains
  - Regarding this, is there any official Palo Alto documentation on this feature? It seems odd to me that a domain or user name would be truncated like this.

Regarding the custom log format change, this is a feature request and I think it would be better to open a separate issue detailing the underlying problem and proposed changes. It would make it easier for us to figure out what is going on and determine from a product perspective if we want to integrate this. You could still open a new PR with the proposed changes, but the main point is to keep it separated from the bug fixes listed above. There will likely be more discussion on this, and I don't want to hold up bug fixes going in.

The following items should be reverted:

- Kibana version upgrade to 8.10.0
- Migrating dashboards

We aren't addressing the issues mentioned by elastic-package build/check quite yet. When we do, we'll tackle them in a separate, self-contained PR (likely as a bulk change PR, since this affects multiple integrations).

marioschaefer commented:
> Thank you for the contribution, @marioschaefer!
> 
> There are a few too many things happening in this PR and I recommend the following to help cut it down in size:
> 
> Let's keep the following changes since that seems to be the main goal of the PR, in particular:
> 
>     * Parsing domains that only contain a '.' and usernames that only contain a ':'
> 
>     * Parsing of truncated usernames and domains
>       
>       * Regarding this, is there any official Palo Alto documentation on this feature? It seems odd to me that a domain or user name would be truncated like this.
> 
> 
> Regarding the custom log format change, this is a feature request and I think it would be better to open a separate issue detailing the underlying problem and proposed changes. It would make it easier for us to figure out what is going on and determine from a product perspective if we want to integrate this. You could still open a new PR with the proposed changes, but the main point is to keep it separated from the bug fixes listed above. There will likely be more discussion on this, and I don't want to hold up bug fixes going in.
> 
> The following items should be reverted:
> 
>     * Kibana version upgrade to 8.10.0
> 
>     * Migrating dashboards
> 
> 
> We aren't addressing the issues mentioned by elastic-package build/check quite yet. When we do, we'll tackle them in a separate, self-contained PR (likely as a bulk change PR, since this affects multiple integrations).

Hi @taylor-swanson,

thank you for you comment.
I understand the part about kibana and will revert the two commits regarding: 

- Kibana version upgrade to 8.10.0
- Migrating dashboards

With regard to the shortened domainnames:
- no i couldn't find any official documentation about that

We have many Palo Altos(>100) and i can see that happening on almost all of them(at least if they are configured to use userids). 
I'm not sure if it's a bug or a feature, but at least it's happening in the wild and the logparser should parse these messages - instead of failing and discarding the logmessage.

Regarding the custom log format change:
I'm torn as to whether its an enhancement or a bugfix, 
Because if i configure the custom-log-format as suggested by palo alto, the log is parsed incorrectly and to (bug)fix this i have to use a different format of the custom-log-format.
In view of these facts, i would classify it as a bugfix.

So question is: do i still have to move this to a new PR or can we leave it here?!

taylor-swanson commented:
> We have many Palo Altos(>100) and i can see that happening on almost all of them(at least if they are configured to use userids).
I'm not sure if it's a bug or a feature, but at least it's happening in the wild and the logparser should parse these messages - instead of failing and discarding the logmessage.

Fair point and I would certainly agree we should not be dropping messages.

> Regarding the custom log format change:
I'm torn as to whether its an enhancement or a bugfix,
Because if i configure the custom-log-format as suggested by palo alto, the log is parsed incorrectly and to (bug)fix this i have to use a different format of the custom-log-format.
In view of these facts, i would classify it as a bugfix.

@jamiehynds, what do you think about this? I haven't heard of the custom log format, but then again I haven't gone looking for it. I'm also not sure if we've had other requests for this yet. The documentation does mention these fields being in custom logs only and is not in the default format:

<img width="676" alt="Screenshot 2024-03-07 at 8 27 42 AM" src="https://github.com/elastic/integrations/assets/90622908/f69512ba-7e88-4778-bb8a-39cb38261927">

To me at least, this suggests that it is an enhancement.

marioschaefer commented:
> > We have many Palo Altos(>100) and i can see that happening on almost all of them(at least if they are configured to use userids).
> > I'm not sure if it's a bug or a feature, but at least it's happening in the wild and the logparser should parse these messages - instead of failing and discarding the logmessage.
> 
> Fair point and I would certainly agree we should not be dropping messages.
> 
> > Regarding the custom log format change:
> > I'm torn as to whether its an enhancement or a bugfix,
> > Because if i configure the custom-log-format as suggested by palo alto, the log is parsed incorrectly and to (bug)fix this i have to use a different format of the custom-log-format.
> > In view of these facts, i would classify it as a bugfix.
> 
> @jamiehynds, what do you think about this? I haven't heard of the custom log format, but then again I haven't gone looking for it. I'm also not sure if we've had other requests for this yet. The documentation does mention these fields being in custom logs only and is not in the default format:
> <img alt="Screenshot 2024-03-07 at 8 27 42 AM" width="676" src="https://private-user-images.githubusercontent.com/90622908/310913318-f69512ba-7e88-4778-bb8a-39cb38261927.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDk4MjczNDEsIm5iZiI6MTcwOTgyNzA0MSwicGF0aCI6Ii85MDYyMjkwOC8zMTA5MTMzMTgtZjY5NTEyYmEtN2U4OC00Nzc4LWJiOGEtMzljYjM4MjYxOTI3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMDclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzA3VDE1NTcyMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTY5MGU3NjMwZTViMGJmZjU4YzkyYzNjODdmNDZjNTY1MzA4NzQzNDM2NzY1NDNlNThmYmI2NjNhMTU2Y2MwOTcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.L1nbL5EpOKKrPUhvfvwYT7AeINlgnWVJ0YCX_jnVWtA">
> 
> To me at least, this suggests that it is an enhancement.

There are even samples of this custom-log-format in the orginial test-samples(line 1-3):
https://github.com/elastic/integrations/blob/127b2a08804fe91a96067deda2e418fd7e4eca65/packages/panw/data_stream/panos/_dev/test/pipeline/test-panw-panos-config-sample.log#L1
But with the original parser(in combination with the log-format suggested by Palo Alto) the logs are parsed wrong.

taylor-swanson commented:
Yeah okay I think I finally understand what's happening here with the custom log format.

The main issue is that the two optional fields are in the middle of the message, and there's not a good way to detect if these fields are present or not. The CSV processor unfortunately doesn't let you conditionally run it based on the number of fields present, it's purely best effort. I also agree that using a split on token in a script processor isn't a good idea (due to the possibility of separators being in a quoted section). You'd basically have to implement a CSV parser just to get the number of fields, then run in it through the correct, real CSV parser.

Now here's the tricky part... what do we do about it. The existing test cases seem to be written with the assumption that custom logs are being used. I don't know why that's the case. Regardless, this has been in and running for the past almost 2 years. A quick look at the issues and I'm not seeing any other bug reports on this. That's not to say that it isn't a bug (it is), it just seems no one has noticed until now. The part that concerns me here is changing expected behavior. If we all of a sudden require a log format change, that's a breaking change to existing deployments.

Let me put it another way, if the typical way to set up these logs is to enable the custom log format, things work normally in the integration. If one were to use the default format, no "errors" would surface, but fields are extracted incorrectly. Either no one is using the default format, or no one has noticed the incorrect field mappings in their deployments. If we were to require a change to the custom format, that would break all existing deployments using the existing format. I'd almost rather go back to trying to figure out how many fields are in the message and selecting the right CSV processor based on that. I'm testing out a script processor for properly counting the number of commas in the message.

taylor-swanson commented:
Creating the script to count the fields wasn't too difficult, and does produce the correct number of fields, even when quotes are in use:

```yaml
  - script:
      description: "Count the number of fields in the message"
      tag: "panw_config_field_count"
      lang: painless
      source: >-
        if (ctx["_temp_"] == null) {
          ctx["_temp_"] = new HashMap();
        }
        
        def count = 1;
        def inQuote = false;
        
        for (int i = 0, n = ctx["message"].length(); i < n; i++) {
          char c = ctx["message"].charAt(i);
          if (c == (char)'"') {
            inQuote = !inQuote;
          }
          if (!inQuote && c == (char)',') {
            count++;
          }
        }
        
        ctx["_temp_"]["field_count"] = count;
```

```
{
      "processor_results": [
        {
          "processor_type": "script",
          "status": "success",
          "doc": {
            "_index": "index",
            "_version": "-3",
            "_id": "id",
            "_source": {
              "message": """1.1.1.1,,set,admin-username,Web,Succeeded, vsys vsys1 address test123,,"test123 { ip-netmask 1.1.1.1/32; tag [ CHECK_RULE dmz-globalprotect dmz1 dmz3 ]; description ""this, is a test. with, two comma""; } ",7304387121517691177,0x0,0,0,0,0,,firewall-hostname,0,,0,2024-02-29T15:33:29.337+01:00""",
              "_temp_": {
                "field_count": 21 <----- HERE
              }
            },
            "_ingest": {
              "pipeline": "_simulate_pipeline",
              "timestamp": "2024-03-07T18:43:08.3126089Z"
            }
          }
        }
      ]
    }
```

The problem I'm running into now is apparently our sample logs dropped the last two fields `FUTURE_USE` and `High Resolution Timestamp`. This means that the standard format can have either 17 or 19 fields and the custom format can have either 19 or 21 fields. If we get a log with 19 fields, which one is it? I'm hoping that it's just our sample data that's bad and not that this is a real scenario (the last two fields being dropped). If it is, I'm not sure how we solve this, there's too much ambiguity.

marioschaefer commented:
> Creating the script to count the fields wasn't too difficult, and does produce the correct number of fields, even when quotes are in use:
> 
> ```yaml
>   - script:
>       description: "Count the number of fields in the message"
>       tag: "panw_config_field_count"
>       lang: painless
>       source: >-
>         if (ctx["_temp_"] == null) {
>           ctx["_temp_"] = new HashMap();
>         }
>         
>         def count = 1;
>         def inQuote = false;
>         
>         for (int i = 0, n = ctx["message"].length(); i < n; i++) {
>           char c = ctx["message"].charAt(i);
>           if (c == (char)'"') {
>             inQuote = !inQuote;
>           }
>           if (!inQuote && c == (char)',') {
>             count++;
>           }
>         }
>         
>         ctx["_temp_"]["field_count"] = count;
> ```
> 
> ```
> {
>       "processor_results": [
>         {
>           "processor_type": "script",
>           "status": "success",
>           "doc": {
>             "_index": "index",
>             "_version": "-3",
>             "_id": "id",
>             "_source": {
>               "message": """1.1.1.1,,set,admin-username,Web,Succeeded, vsys vsys1 address test123,,"test123 { ip-netmask 1.1.1.1/32; tag [ CHECK_RULE dmz-globalprotect dmz1 dmz3 ]; description ""this, is a test. with, two comma""; } ",7304387121517691177,0x0,0,0,0,0,,firewall-hostname,0,,0,2024-02-29T15:33:29.337+01:00""",
>               "_temp_": {
>                 "field_count": 21 <----- HERE
>               }
>             },
>             "_ingest": {
>               "pipeline": "_simulate_pipeline",
>               "timestamp": "2024-03-07T18:43:08.3126089Z"
>             }
>           }
>         }
>       ]
>     }
> ```
> 
> The problem I'm running into now is apparently our sample logs dropped the last two fields `FUTURE_USE` and `High Resolution Timestamp`. This means that the standard format can have either 17 or 19 fields and the custom format can have either 19 or 21 fields. If we get a log with 19 fields, which one is it? I'm hoping that it's just our sample data that's bad and not that this is a real scenario (the last two fields being dropped). If it is, I'm not sure how we solve this, there's too much ambiguity.
> 
> EDIT: The `FUTURE_USE` and `High Resolution Timestamp` fields were added in PAN-OS 10.1. Our sample logs were generated on PAN-OS 10.0 or earlier. My comment on ambiguity still stands.

Well well well, i guess that's a good solution on counting the fields.

Regarding the other problem:
The timestamp was introduces with 10.0(`The High Resolution Timestamp is supported for logs received from managed firewalls running PAN-OS 10.0 and later releases.`) - the PAN Documentation is a bit misleading since this sentence is not in the 10.0 but in the 10.1 docs. But 10.0 is EOL anyways.
We could also check if the last field is a Timestamp in the given format(with regex), to find out if its the new or old logformat.
Or drop the old logformat completely, since PAN-OS 9.1 is almost at its EOL([30.06.2024 in general](https://www.paloaltonetworks.com/services/support/end-of-life-announcements/end-of-life-summary#pan-os-panorama) resp. [31.10.2024 for PA-3000 series](https://www.paloaltonetworks.com/services/support/end-of-life-announcements/hardware-end-of-life-dates.html)).

jamiehynds commented:
@taylor-swanson I'd be wary of dropping support for v9.1 given that there's still ~3 months before end of life. Although most folks have migrated or migrating before end of life, I'd ideally like to avoid catching anyone off guard if we drop support prior to the EOL date. However, once it hits EOL there's no issues on dropping support on our end.

marioschaefer commented:
I see the point, on the other hand: People who are still using 9.1 can just stick to the integration-version they are using now - nothing will change for them. they seem to be people who don't like to update anyway 😄 
  
but people who are using a modern software release can upgrade to the new version and get a working logparser. on our side its more easy as we don't have to do "magic" to parse logformats from 1000 BC.
We could raise the Integration-Version to 4.0.0 to make it clear that there is a "bigger" change and explicitly mention that the support for 9.1 logformat is dropped.

marioschaefer commented:
@taylor-swanson alright, i've added your commit to my PR and also fixed the readme. i will test this on monday in our production environment.

marioschaefer commented:
@taylor-swanson Your solution seems to be working, thanks for the help!  
I just have one more point: can we reenable the palo high precision timestamp? as this is also introduced with 10.0, i would expect this to be also fixed in this PR. As the CSV-processor doesn't care if there are too many or too less fields, we can just add the two needed fields to the csv parser. If they are there, they get parsed, if not - not.

taylor-swanson commented:
@marioschaefer,

Great to hear!

Yes, let's add the high res timestamp. For whatever reason, it doesn't look like it was added in the first place for config logs. I see there's already a processor to handle it in the default pipeline, so what you have is good.

taylor-swanson commented on packages/panw/_dev/build/docs/README.md:
```diff
@@ -23,6 +25,8 @@ To configure syslog monitoring, please follow the steps mentioned in the [_Confi
   - It can be found under Advanced Options and can be configured as per requirements. The default value of `max_message_size` is set to 50KiB.
 
 - If the TCP input is used, it is recommended that PAN-OS is configured to send syslog messages using the IETF (RFC 5424) format. In addition, RFC 6587 framing (Octet Counting) will be enabled by default on the TCP input.
+- If you want to see the configuration before and after the change(fields `before-change-detail` and `after-change-detail`) in the [config-log](https://docs.paloaltonetworks.com/pan-os/11-1/pan-os-admin/monitoring/use-syslog-for-monitoring/syslog-field-descriptions/config-log-fields), please use the following [custom log format in the syslog server profile](https://docs.paloaltonetworks.com/pan-o ...
+  ``1,$receive_time,$serial,$type,$subtype,2561,$time_generated,$host,$vsys,$cmd,$admin,$client,$result,$path,$before-change-detail,$after-change-detail,$seqno,$actionflags,$dg_hier_level_1,$dg_hier_level_2,$dg_hier_level_3,$dg_hier_level_4,$vsys_name,$device_name,$dg_id,$comment,0,$high_res_timestamp``
```
I think this can be removed now? We won't require any format changes.

marioschaefer replied:
I think we should keep that. Since if you activate the "custom log format" in the Palo Alto GUI, you have to enter the "custom log format" you want to use. Since its "custom" there is no default in the Palo GUI, only the example in the Palo docs. Long story short: this "custom log format" is a good starting point, and the only working log format.

taylor-swanson replied:
Ah gotcha. I agree, then. Let's keep it.

taylor-swanson commented on packages/panw/data_stream/panos/_dev/test/pipeline/test-panw-panos-config-sample.log-expected.json:
```diff
@@ -289,6 +301,73 @@
+                    "device_group_hierarchy3": "0",
+                    "device_group_hierarchy4": "0",
+                    "device_group_id": "0",
+                    "path": "vsys vsys1 address test123",
```
Shouldn't we be seeing a high resolution timestamp here? Perhaps we need to re-generate the expected files?

marioschaefer replied:
oh, yes - you are right. it's missing in the `expected.json`

taylor-swanson commented on packages/panw/changelog.yml:
```diff
@@ -1,4 +1,9 @@
 # newer versions go on top
+- version: "3.22.1"
```
Since this is a bigger change, let's use `3.23.0` as the version number.

marioschaefer replied:
you already pushed 3.23.1 in #9333 
i changed my version to `3.24.0`

taylor-swanson commented:
/test

elasticmachine commented:
## :broken_heart: Build Failed
* [Buildkite Build](https://buildkite.com/elastic/integrations/builds/9830)
* Commit: 59aea7705b910199ae5bc546cf79a013eb9de662

### Failed CI Steps
* [:pipeline::arrow_up: Upload Pipeline: .buildkite/pipeline.yml](https://buildkite.com/elastic/integrations/builds/9830#018e75dd-41ff-4004-8e21-e84141814d76)

<!--buildkite-pr-comment-integrations
{"builds":[{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/integrations/builds/9830","number":9830,"commit":"59aea7705b910199ae5bc546cf79a013eb9de662"}],"number":9830}
buildkite-pr-comment-->

taylor-swanson commented:
Also looks like you'll need to resolve the conflicts before tests can run.

marioschaefer commented on packages/panw/data_stream/panos/elasticsearch/ingest_pipeline/default.yml:
```diff
@@ -104,24 +104,26 @@ processors:
       pattern_definitions:
-        USERNAME: '[ a-zA-Z0-9._-]+[$]?'
+        USERNAME: '[ a-zA-Z0-9.:_-]+[$]?'
+        PATRUNKATEDHOSTNAME: '(?:\.{0,1}|\b(?:[0-9A-Za-z][0-9A-Za-z-]{0,62})(?:\.{1,2}(?:[0-9A-Za-z][0-9A-Za-z-]{0,62}))*(\.?|\b))'
```
This pattern is from PR #8780, but i will add a matching test case.

marioschaefer replied:
ah, i already added a test case for this pattern: 
https://github.com/elastic/integrations/pull/9247/files#diff-26f926244f9bec70b3348b13e3556af8def2e05390d4dca68088b5bb889d587cR196
and expected: https://github.com/elastic/integrations/pull/9247/files#diff-099d0ab80feebc4ddf2002da70283bdc9193e21e393b62d50d7d1bf658493cbeR32603

taylor-swanson replied:
Perfect, thanks!

marioschaefer commented:
> Also looks like you'll need to resolve the conflicts before tests can run.

@taylor-swanson fixed

taylor-swanson commented:
/test

elasticmachine commented:
### :rocket: Benchmarks report




To see the full report comment with `/test benchmark fullreport`
<!--COMMENT_GENERATED_WITH_ID_benchmark-report-->

elasticmachine commented:
## :green_heart: Build Succeeded
* [Buildkite Build](https://buildkite.com/elastic/integrations/builds/9835)
* Commit: 3ec15a6f19b6bea30b2c6f6e9b1318cc9b783d13

### History
* :broken_heart: [Build #9830](https://buildkite.com/elastic/integrations/builds/9830) failed 59aea7705b910199ae5bc546cf79a013eb9de662

<!--buildkite-pr-comment-integrations
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/integrations/builds/9835","number":9835,"commit":"3ec15a6f19b6bea30b2c6f6e9b1318cc9b783d13"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/integrations/builds/9830","number":9830,"commit":"59aea7705b910199ae5bc546cf79a013eb9de662"}],"number":9835}
buildkite-pr-comment-->

elasticmachine commented:
Package panw - 3.24.0 containing this change is available at https://epr.elastic.co/search?package=panw
